R/chat_app.R

```diff
@@ -0,0 +1,216 @@
+#'   chat_mod_server("openai", openai)
+#' }
+#'
+#' shinyApp(ui, server)
+#' }
+#'
+#' @param client A chat object created by \pkg{ellmer}, e.g.
+#'   [ellmer::chat_openai()] and friends.
+#' @param ... In `chat_app()`, additional arguments are passed to
+#'   [shiny::shinyApp()]. In `chat_mod_ui()`, additional arguments are passed to
+#'   [chat_ui()].
+#'
+#' @returns
+#'   * `chat_app()` returns a [shiny::shinyApp()] object.
+#'   * `chat_mod_ui()` returns the UI for a shinychat module.
+#'   * `chat_mod_server()` includes the shinychat module server logic, and
+#'     and returns the last turn upon successful chat completion.
+#'
+#' @describeIn chat_app A simple Shiny app for live chatting.
+#' @export
+chat_app <- function(client, ...) {
+  check_ellmer_chat(client)
+
+  ui <- bslib::page_fillable(
+    chat_mod_ui("chat", client = client, height = "100%"),
+    shiny::actionButton(
+      "close_btn",
+      label = "",
+      class = "btn-close",
+      style = "position: fixed; top: 6px; right: 6px;"
+    )
+  )
+
+  server <- function(input, output, session) {
+    chat_mod_server("chat", client)
+
+    shiny::observeEvent(input$close_btn, {
+      shiny::stopApp()
+    })
+  }
+
+  shiny::shinyApp(ui, server, ...)
+}
+
+check_ellmer_chat <- function(client) {
+  if (!inherits(client, "Chat")) {
+    abort("`client` must be an `ellmer::Chat` object.")
+  }
+}
+
+#' @describeIn chat_app A simple chat app module UI.
+#' @param id The chat module ID.
+#' @param messages Initial messages shown in the chat, used when `client` is not
+#'   provided or when the chat `client` doesn't already contain turns. Passed to
+#'   `messages` in [chat_ui()].
+#' @export
+chat_mod_ui <- function(id, ..., client = NULL, messages = NULL) {
+  if (!is.null(client)) {
+    check_ellmer_chat(client)
+
+    client_msgs <- map(client$get_turns(), function(turn) {
+      content <- ellmer::contents_markdown(turn)
+      if (is.null(content) || identical(content, "")) {
+        return(NULL)
+      }
+      list(role = turn@role, content = content)
+    })
+    client_msgs <- compact(client_msgs)
+
+    if (length(client_msgs)) {
+      if (!is.null(messages)) {
+        warn(
+          "`client` was provided and has initial messages, `messages` will be ignored."
+        )
+      }
+      messages <- client_msgs
+    }
+  }
+
+  shinychat::chat_ui(
+    shiny::NS(id, "chat"),
+    messages = messages,
+    ...
+  )
+}
+
+#' @describeIn chat_app A simple chat app module server.
+#' @export
+chat_mod_server <- function(id, client) {
+  check_ellmer_chat(client)
+
+  append_stream_task <- shiny::ExtendedTask$new(
+    function(client, ui_id, user_input) {
+      promises::promise(function(resolve, reject) {
+        tryCatch(
+          resolve(client$stream_async(user_input)),
+          error = function(err) {
+            reject(err)
+          }
+        )
+      })$then(function(stream) {
+        chat_append(ui_id, stream)
+      })
```

Comment:
```suggestion
      promises::then(client$stream_async(user_input), function(stream) {
        chat_append(ui_id, stream)
      })
```

Reply:
Will this catch errors that happen before `$stream_async()` returns a promise?

I don't think so -- I had something quite similar initially and needed this pattern to ensure we caught very early errors in the `client$stream_async()` call.

Reply:
Yeah, the suggested code gives me

```
Caused by error in `as.promise.default()`:
! Don't know how to convert object of class coro_generator_instance into a promise
```

Actually, if everything works as expected you just need

```
stream <- client$stream_async(user_input)
chat_append(ui_id, stream)
```

inside the extended task, but you get into trouble when `$stream_async()` fails.

Reply:
Oh, good point. That's unfortunate, can we work harder to catch those?

In the meantime, I believe `promise_resolve` catches the synchronous error. But it's also fine with me if you want to leave it as-is.

```suggestion
      promises::then(promises::promise_resolve(client$stream_async(user_input)), function(stream) {
        chat_append(ui_id, stream)
      })
```

Reply:
Perfect, thanks!